cmd/snap-update-ns/bootstrap.c

```diff
@@ -15,12 +15,20 @@
  *
```

Comment:
This comment is for all of bootstrap.c since github doesn't allow contextual comments for non-patched areas.

= in bootstrap(), line 254 =
```
    // We don't have argc/argv so let's imitate that by reading cmdline
    char cmdline[1024];
    memset(cmdline, 0, sizeof cmdline);
```

The memset() could be optimized away. Instead, just skip the memset() and do:
```
    char cmdline[1024] = { 0 };
```

= in read_cmdline(), line 54 =
```
    memset(buf, 0, buf_size);
```

The memset() could be optimized away. Instead, do:
```
   // Ensure buf is initialized to all NULLs since our parsing of cmdline
   // with its embedded NULLs depends on this. Use for loop instead of memset()
   // to guarantee this initialization won't be optimized away.
   for (int i=0; i<buf_size; i++)
       buf[i] = '\0';
```

(dealing with strings with embedded NULLs in C are annoying)

= in read_cmdline(), line 59 =
```
    } else if (num_read == buf_size) {
```

This is an off by 1. `num_read == buf_size` is ok because 'foo@bar@' will have num_read as 8 and if buf_size was also 8, it would fit. You want this instead:
```
    } else if (num_read > buf_size) {
        ...
    } else if (buf[num_read-1] != '\0') {
        // the kernel should always make this NULL terminated, but let's guard
        // against kernel bugs.
        bootstrap_errno = 0;
        bootstrap_msg = "short read on /proc/self/cmdline";
        return -1;
    }
```

= in bootstrap(), line 261 =
```
    // Find the name of the called program. If it is ending with "-test" then do nothing.
    // NOTE: This lets us use cgo/go to write tests without running the bulk
    // of the code automatically. In snapd we can just set the required
    // environment variable.
```
What environment variable? Is this variable used by snap-confine? If so, what would prevent someone from doing: `WHATEVER_THE_VARIABLE_IS=/home/bad/foo.test snap-confine ...`

```
    const char* argv0 = find_argv0(cmdline, (size_t)num_read);
    if (argv0 == NULL) {
        bootstrap_errno = 0;
        bootstrap_msg = "argv0 is corrupted";
        return;
    }
    const char* argv0_suffix_maybe = strstr(argv0, ".test");
    if (argv0_suffix_maybe != NULL && argv0_suffix_maybe[strlen(".test")] == '\0') {
        bootstrap_errno = 0;
        bootstrap_msg = "bootstrap is not enabled while testing";
        return;
    }
```

Also, the comment says "-test" but you are checking for ".test". Regardless, hardcoding this makes me uneasy because if the env variable handling gets messed up, we potentially open ourselves up to abuse. Since this is only for test code, maybe wrap this in a #define that you set as part of the build?


= in find_argv0(), line 68 =
```
    // cmdline is an array of NUL ('\0') separated strings.
    size_t argv0_len = strnlen(buf, num_read);
    if (argv0_len == num_read) {
        // ensure that the buffer is properly terminated.
        return NULL;
    }
    return buf;
```

We've already guaranteed in read_cmdline() that the string is NULL-terminated. I suggest simply changing this to:
```
    // cmdline is an array of NUL ('\0') separated strings and guarnteed to
    // be NULL-terminated via read_cmdline().
    return buf;
```

Of course, that makes the function rather pointless, so you could just use 'buf' where you call find_argv0() and put your comment there about how buf contains strings separated by NULLs and is guaranteed to be NULL-terminated.

= in bootstrap(), line 265, 280, 296 =
The casts from ssize_t to size_t are fine since you are casting from signed to unsigned and we verified num_read is always >= 0.

= in bootstrap(), line 281 =
```
    // Find the name of the snap by scanning the cmdline.  If there's no snap
    // name given, just bail out. The go parts will scan this too.
    const char* snap_name = find_snap_name(cmdline, (size_t)num_read);
    if (snap_name == NULL) {
        return;
    }
```

We're supposed to be using bootstrap.c for bulletproof input validation. As such, let's not bail out but instead set bootstrap_msg. Eg:
```
    const char* snap_name = find_snap_name(cmdline, (size_t)num_read);
    if (snap_name == NULL) {
        bootstrap_errno = 0;
        bootstrap_msg = "snap name not found in argv1";
        return;
    }
```

= in find_snap_name(), line 94 =
```
    // Skip the zeroth argument as well as any options.
    do {
        size_t arg_len = strnlen(buf, num_read);
        if (arg_len + 1 >= num_read) {
            return NULL;
        }
        num_read -= arg_len + 1;
        buf += arg_len + 1;
    } while (buf[0] == '-');

    char* snap_name = buf;
    if (*snap_name == '\0') {
        return NULL;
    }
```

This is more complicated than it has to be. We guarantee in read_cmdline() that buf will be NULL-terminated, so you can use strlen() and skip all the num_read changes. Ie:
```
    // Skip the zeroth argument as well as any options. We know buf is NULL
    // padded and terminated from read_cmdline().
    do {
        size_t arg_len = strlen(buf);
        buf += arg_len + 1;
    } while (buf[0] == '-');

    char *snap_name = buf;
    if (strlen(snap_name) == 0) {
        return NULL;
    }
```

= in validate_snap_name(), skip_lowercase_letters(), skip_digits() and skip_one_char () =
This is very nearly a full code copy of what is in cmd/libsnap-confine-private/snap.c. Ideally we'd be sharing the code since bootstrap.c and snap-confine are doing the same thing for the same reasons (ie trying to run safely with privileges).

I think adding a TODO above each of these functions that says something along the lines of:
```
// TODO: use functions from cmd/libsnap-confine-private/snap.c instead
```

would be a good idea. Actually performing that work is not a blocker for this PR, but please consider it for a future improvement.

= in bootstrap(), line 295 =
```
    // When we are running under "--from-snap-confine" option skip the setns
    // call as snap-confine has already placed us in the right namespace.
    const char* option = find_1st_option(cmdline, (size_t)num_read);
    if (option != NULL && strncmp(option, "--from-snap-confine", strlen("--from-snap-confine")) == 0) {
        return;
    }
```

In the spirit of bulletproofing, we should error if get an unsupported option. I suggest something like:
```
    const char* option = find_1st_option(cmdline, (size_t)num_read);
    if (option != NULL) {
        // When we are running under "--from-snap-confine" option skip the
	// setns call as snap-confine has already placed us in the right
        // namespace.
        if (strncmp(option, "--from-snap-confine", strlen("--from-snap-confine")) == 0) {
            return;
        } else {
            bootstrap_errno = 0;
            bootstrap_msg = "unsupported option";
            return;
        }
    }
```

= in find_1st_option(), line 111 =
```
    size_t argv0_len = strnlen(buf, num_read);
    if (argv0_len + 1 >= num_read) {
        return NULL;
    }
    size_t pos = argv0_len + 1;
    if (buf[pos] == '-') {
        return &buf[pos];
    }
    return NULL;
```

buf is guarnteed to be NULL-terminated, so this is more complicated than it has to be. I suggest simply:
```
    size_t pos = strlen(buf) + 1;
    if (buf[pos] == '-') {
        return &buf[pos];
    }
    return NULL;
```

= in setns_into_snap(), line 125 =
Please initialize buf as:
```
char buf[PATH_MAX] = { 0 };
```

Also, in cmd/snap-confine/ns-support.c, you set SC_NS_DIR and SC_NS_MNT_FILE , but you hardcode it here. Why?

Also, snprintf doesn't set errno here:
```
    int n = snprintf(buf, sizeof buf, "/run/snapd/ns/%s.mnt", snap_name);
    if (n >= sizeof buf || n < 0) {
        bootstrap_errno = errno;
```

Please just use `bootstrap_errno = 0;`

Finally, the logic in cmd/snap-confine/ns-support.c sc_create_or_join_ns_group() is very different than what is in setns-into_snap():
- you use O_NOFOLLOW in snap-confine, but not here
- you don't use fstatfs() here
- nothing about NSFS_MAGIC or PROC_SUPER_MAGIC
- you aren't performing a chdir()

Can you comment in more detail why in snap-confine this was needed but here it
is not?